test/AutoDiff/differentiable_func_type_parse.swift

```diff
@@ -0,0 +1,32 @@
+// RUN: %target-swift-frontend -parse -verify %s
+
+let a: @differentiable (Float) -> Float // okay
+
+let b: @differentiable(linear) (Float) -> Float // okay
+
+// Generic type test.
+struct A<T: Differentiable> {
+  func foo() {
+    let _: @differentiable(linear) (T) -> T // okay
+  }
+}
+
+// expected-error @+1 {{expected ')' after 'linear' in '@differentiable' attribute}}
+let c: @differentiable(linear (Float) -> Float
+
+// expected-error @+1 {{unexpected argument 'notValidArg' in '@differentiable' attribute}}
+let c: @differentiable(notValidArg) (Float) -> Float
+
+// Using 'linear' as a type
+struct A {
+  struct linear : Differentiable {}
+  let property: @differentiable (linear) -> Float // okay
+  let property: @differentiable(linear) (linear) -> linear // okay
+  let property: @differentiable (linear, linear) -> linear // okay
+  let property: @differentiable (linear, Float) -> linear // okay
+  let property: @differentiable (Float, linear) -> linear // okay
+  let property: @differentiable(linear) (linear, linear, Float, linear)
+    -> Float // okay
+  // expected-error @+1 {{expected ')' after 'linear' in '@differentiable' attribute}}
+  let property: @differentiable(linear (linear) -> Float
```

Comment:
I'll throw an unlikely case here... Feel free to ignore it.
```swift
typealias linear = (Float) -> Float
let _: @differentiable(linear) // hell yeah
```

Reply:
Ahh yes with the typealias I'll add that!

Reply:
Good catch! As this test made me realize that this works like expected:
```swift
let property: @differentiable linear // okay
```
However, this should work but doesn't:
```swift 
let property: @differentiable(linear) linear // should be okay, but fails
```

```
error: unexpected error produced: consecutive declarations on a line must be separated by ';'
  let property: @differentiable(linear) linear // okay
                                       ^
```

Reply:
Offline we decided to comment out this test case and file a [JIRA ticket](https://bugs.swift.org/browse/TF-576) about it.